### ISSUE TITLE
configuration of clangd?

### ISSUE THREAD

**Issue by mcepl** [prabirshrestha/vim-lsp#67]
When trying to autocomplete string ``goa_account_`` in [t_kinit.c](https://github.com/prabirshrestha/vim-lsp/files/1540736/t_kinit.c.txt) (line 89) I enter
```c
if (!goa_account
```
and I press ``Ctrl-X-Ctrl-O``, but the suggested autocompletion is completely wrong:

![screenshot from 2017-12-07 22-36-30](https://user-images.githubusercontent.com/198999/33740587-7edb2646-dba1-11e7-99bc-84809ce1c9b3.png)

[vim-lsp-log.txt](https://github.com/prabirshrestha/vim-lsp/files/1540744/vim-lsp-log.txt) is record of the conversation between vim and clangd.

When trying the same in vscode (not sure, whether it uses the same clangd or something of its own), autocompletion works just fine.

All that on RHEL-7, with clangd 5.0.0, vim 8.0.1367, vscode 1.18.1, vim-lsp from e7ab592.

Could the problem be that clangd is missing some configuration. When using Intellisense in VSCode, I need to add rather large configuration (especially include directories) in [c_cpp_properties.json](https://github.com/prabirshrestha/vim-lsp/files/1540764/c_cpp_properties.json.txt). Isn't the problem of vim-lsp, that it doesn't allow such configuration?

**Comment by prabirshrestha** [comment]
@mcepl did you try other small projects?

Given that I use other lang server every day I haven't noticed this so I think it is most likely not setting the configuration properly. VsCode most likely has some good default so it just works.

**Comment by mcepl** [comment]
OK, made a special new directory and wrote this file (``hello.c``):
```c
#include <stdio.h>

void main() {
    pr
}
```
With cursor on ``pr`` I have pressed ``Ctrl-X,Ctrl-O`` and this is the result:
![screenshot from 2018-01-31 22-06-43](https://user-images.githubusercontent.com/198999/35647611-5df82ea8-06d3-11e8-9a12-31e884469170.png)

Also I have an error message: ``use of undeclared identifier 'pr' [clang-diagnostic-error]`` which is not very helpful. I don't think I have to declare ``printf``, do I?

And this is [vim-lsp.log](https://github.com/prabirshrestha/vim-lsp/files/1683341/vim-lsp.log.txt)

**Comment by mcepl** [comment]
BTW, exactly the same results with neovim 0.2.2 (just to be sure, it is not vim8).

**Comment by drodil** [comment]
Same thing here. Not even std functions can be found and lsp log is just giving output "use of undeclared identifier 'std'". I have clangd-6.0.0 so is this supported or should it be version 5? I would love the plugin as it's lightweight compared to for example YouCompleteMe but because of these problems I might change to something else.

**Comment by prabirshrestha** [comment]
Can you create a sample repo that I can try?

Also paste the log.

**Comment by drodil** [comment]
Tried to make a sample repo and it worked there with just three files: main.cpp, inc/class.h and src/class.src. I think this is related to the project configuration I am working on somehow as it's much more complex and is separated in different libraries that use each other. 

How does the clangd server scan for the header files? And is there someway to configure it? I might be forced to use the ctags after all together with https://github.com/yami-beta/asyncomplete-omni.vim

**Comment by drodil** [comment]
It's related to includes, thats for sure. We have the following directory structure:
```
impl
-- inc
---- namespace1 
------ namespace2
-------- test_class.h
-- src
----- namespace1
------- namespace2
--------- test_class.cpp
```

Where test_class.cpp uses:

```cpp
#include "namespace1/namespace2/test_class.hpp
```

It works fine with CMake as the include directories are set there but the clangd does not understand this structure. Changing the include to relative works fine:

```cpp
#include "../../../inc/namespace1/namespace2/test_class.hpp
```
But it's not very beautiful in the source code if you know what I mean. 

Are there any ways to tell the clangd which include directories to use?

**Comment by mcepl** [comment]
Well, ehm, isn't this just a bug in clangd?

**Comment by drodil** [comment]
Yes it seems so. I might try with the latest version 7 if it helps at all.

**Comment by drodil** [comment]
Hah, with clangd-7 it works just fine. Thanks for pointing this out!

**Comment by renstrom** [comment]
It might be worth trying [`cquery`](https://github.com/cquery-project/cquery) instead of `clangd` as well.

https://github.com/prabirshrestha/vim-lsp/wiki/Servers-cquery

**Comment by drodil** [comment]
Sure it's possible but I really don't like adding extra configuration files for things that should work without. When there is extra conf files it's always annoying to keep them updated in case new include folders or build flags are introduced. Ultimate solution would be for the cquery to gather include folders from existing CMake/Make/whatever files.

**Comment by mcepl** [comment]
> #107

Confused? That's about ``cquery``, right? This ticket is about ``clangd``.

**Comment by prabirshrestha** [comment]
@mcepl It is about url decoding which solves for cquery. I'm was wondering if clang has the same issue.

Basically `cquery` returns encoded path `file:///usr/include/c%2B%2B/7.2.0/bits/stringfwd.h` and vim-lsp doesn't know how to handle it. So nothing works. I'm wondering if clang does the same too. 
vim-lsp is currently expecting the path to be `file:///usr/include/c++/7.2.0/bits/stringfwd.h`

**Comment by drodil** [comment]
@prabirshrestha no the problem is not about encoding with clang. But I think it doesn't hurt and try to decode the path. This bug must have been with clangd versions 7< because with the latest it works like a charm.

**Comment by pdavydov108** [comment]
@drodil `compile_commands.json` is required for both cquery and clangd, so cquery requires no more configuration then clangd. In my understanding, it's almost impossible to guess project directory structure in c++ without this json file, cause there is no standard way to do it. Even if it was somehow possible, there are still issues with compiler flags, macros, include dirs, etc. Also, note that cquery has much more features then clangd. I would like to go with clangd myself, cause it looks more mainstream, but just check the list of supported features: [clangd](https://clang.llvm.org/extra/clangd.html#id6) vs [cquery](https://github.com/cquery-project/cquery). By the way, at least with cmake it is trivial to make it generate `compile_commands.json`, just pass `-DCMAKE_EXPORT_COMPILE_COMMANDS=On` to cmake execution command.

**Comment by pdavydov108** [comment]
@mcepl I checked your simple example where you create a new dir with main function. I've added a simple compile_commands.json file to the same dir that looks like so:
```json
[
    {
        "directory": "/home/pablo/exp",
        "command": "/usr/bin/clang++  -g -o main.c.o -c /home/pablo/exp/main.c",
        "file": "/home/pablo/exp/main.c"
    }
]
```
With this file in the root, when I type `pr`, `printf` function with signature is suggested, which is what we want. Tried it with clangd-6.0.

**Comment by mcepl** [comment]
So, I have added this [compile_commands.json](https://github.com/prabirshrestha/vim-lsp/files/1802831/compile_commands.json.txt) to the directory where ``t_kinit.c`` was and then I wrote ``pr`` and pressed ``Ctrl-X,Ctrl-O``. vim responded with ``Pattern not found``. [vim-lsp.log](https://github.com/prabirshrestha/vim-lsp/files/1802835/vim-lsp.log.txt)

**Comment by pdavydov108** [comment]
It is better to use your build system to generate compile_commands.json. I've wrote before how to generate it for cmake, please check how to do it with your build system. In your case there are messages from cquery server 'goa/goa.h file not found', and other similar errors for glib and goabackend. This is because cquery doesn't know where to look for this headers. So, types and functions from this headers never get loaded. Thus, all your compilation unit becomes hard to parse. As a quick fix, you can just add `-I/path/to/goa/goa.h` options for all headers that you have to "command" line in your config. Can you see this errors with signs enabled like this `let g:lsp_signs_enabled = 1`?

**Comment by mcepl** [comment]
I didn't have any build system. This is just a simple C test case for some large project. I just run
```sh
gcc goa-renew-kerberos.c $(pkg-config --cflags --libs goa-1.0 goa-backend-1.0 glib-2.0)
```

**Comment by pdavydov108** [comment]
I see. So maybe you can just get the output of `$(pkg-config --cflags --libs goa-1.0 goa-backend-1.0 glib-2.0)` and then set command like this: `"command": "gcc t_kinit.c <result of pkg config>"` in the compile_commands.json?

**Comment by mcepl** [comment]
OK, [compile_commands.json](https://github.com/prabirshrestha/vim-lsp/files/1803478/compile_commands.json.txt) looks a bit extravagant, but still it doesn't make much difference. ``pri`` still doesn't complete. [vim-lsp.log](https://github.com/prabirshrestha/vim-lsp/files/1803480/vim-lsp.log.txt)

**Comment by prabirshrestha** [comment]
Seems like the server is crashing.

<---(stderr)", 12, "cquery", ["\u001B[0m\u001B[1m\u001B[91m2018-03-12 16:45:47.005 (   2.044s) [stdin        ]            lsp.cc:130    ERR| Failed to read JsonRpc input; exiting\u001B[0m", "\u001B[0m\u001B[1m\u001B[91m", "Loguru caught a signal: SIGTERM", "\u001B[0m"]]
Mon 12 Mar 2018 16:45:47 CET:["<---(stderr)", 12, "cquery", ["\u001B[0m\u001B[1m\u001B[91mStack trace:", "12            0x6afb5d cquery() [0x6afb5d]", "11      0x7f97e38e13d5 __libc_start_main + 245", "10            0x719174 main + 4964", "9             0x717db8 LanguageServerMain(std::string const&, Config*, MultiQueueWaiter*, MultiQueueWaiter*, MultiQueueWaiter*) + 152", "8             0x7176e6 RunQueryDbThread(std::string const&, Config*, MultiQueueWaiter*, MultiQueueWaiter*) + 1798", "7             0x71bce7 void MultiQueueWaiter::Wait<ThreadedQueue<Index_OnIndexed>*, ThreadedQueue<std::unique_ptr<BaseIpcMessage, std::default_delete<BaseIpcMessage> > >*, ThreadedQueue<Index_DoIdMap>*>(ThreadedQueue<Index_OnIndexed>*, ThreadedQueue<std::unique_ptr<BaseIpcMessage, std::default_delete<BaseIpcMessage> > >*, ThreadedQueue<Index_DoIdMap>*) + 151", "6             0x72e8be void std::condition_variable_any::wait<MultiQueueLock<ThreadedQueue<Index_OnIndexed>*, ThreadedQueue<std::unique_ptr<BaseIpcMessage, std::default_delete<BaseIpcMessage> > >*, ThreadedQueue<Index_DoIdMap>*> >(MultiQueueLock<ThreadedQueue<Index_OnIndexed>*, ThreadedQueue<std::unique_ptr<BaseIpcMessage, std::default_delete<BaseIpcMessage> > >*, ThreadedQueue<Index_DoIdMap>*>&) + 110", "5       0x7f97e425582c std::condition_variable::wait(std::unique_lock<std::mutex>&) + 12", "4       0x7f97e8144945 pthread_cond_wait + 197", "3       0x7f97e8148680 /lib64/libpthread.so.0(+0xf680) [0x7f97e8148680]", "2       0x7f97e5ccf215 /home/matej/archiv/knihovna/repos/cquery/build/debug/bin/../lib/clang+llvm-5.0.1-x86_64-linux-gnu-ubuntu-14.04/lib/libclang.so.5(+0x1824215) [0x7f97e5ccf215]", "1       0x7f97e814854b raise + 43", "0       0x7f97e8148680 /lib64/libpthread.so.0(+0xf680) [0x7f97e8148680]\u001B[0m", "\u001B[0m\u001B[1m\u001B[91m2018-03-12 16:45:47.005 (   2.044s) [querydb      ]                  :0     FATL| Signal: SIGTERM\u001B[0m", ""]]
Mon 12 Mar 2018 16:45:47 CET:["s:on_exit", 12, "cquery", "exited", 0]

**Comment by mcepl** [comment]
cquery-project/cquery#510

**Comment by drodil** [comment]
@pdavydov108 I tried it with clangd version 7 without the compile_commands.json file and it worked perfectly. No idea why but going with this for now :)

**Comment by MaskRay** [comment]
There also should be some knob to define the project root (`rootUri` in LSP) https://github.com/cquery-project/cquery/wiki/Emacs#projects-with-subprojects

**Comment by mcepl** [comment]
OK, ``clangd`` is just the opposite: where ``cquery`` generates too little, it generates too much 
[vim-lsp.log.gz](https://github.com/prabirshrestha/vim-lsp/files/1804723/vim-lsp.log.gz) and still nonsensical completion 
![screenshot from 2018-03-12 22-47-22](https://user-images.githubusercontent.com/198999/37311893-5bedd6e0-2649-11e8-82eb-ba7aebe3b2b8.png). Anyway, I don't working with ``C`` that much.
